### PR TITLE
refactor(ios/engine): package-version query caching + package state properties

### PR DESCRIPTION
**Pull request by jahorton**
This PR gives the `KeymanPackage` class `public` ownership of update query checks and adds new properties regarding the state of a package.  Furthermore, it (finally) adds support for _caching_ the result of query checks so that we can throttle the app's frequency of such checks once the update code is reworked to utilize it.  That will reduce the amount of data needed by the app.

It also adds serious post-processing for package-version checks designed to facilitate background notifications for resource updates.  We're not doing that now, but this should remove a serious hurdle that would stand in the way of implementing the feature - in the app's top level UI, via system notification... or possibly both.  (At some point in the future.)

This also provides a mechanism to detect and track deprecations as part of package update checks, although our API doesn't return this information quite yet.  Ideally, we'll eventually be able to prompt users whenever a keyboard is deprecated and make targeted recommendations.  (After all, we do track a `deprecatedBy` field in our backing database.)

Finally, this PR adds a lot of testing infrastructure to detect any stability issues in the reworked code, helping improve the app's stability.

----

A number of side-effects will be altered in upcoming PRs centered upon refactoring iOS's update paradigm.  While I'd like to lump those in here... well, I'm already hitting 500+ lines as it is.  Granted, much of that is testing, but I'd still prefer to keep the per-PR scope _relatively_ controlled.

It's for similar reasons that I'm not instantly fixing a lot of the new deprecation warnings that result.  Many of them are in update code that will be changed shortly, with most others located in the keyboard installation process that will be ripped out.

### CODE REVIEW

File: ios/engine/KMEI/KeymanEngine/Classes/KeymanPackage.swift [hunk 1]
```diff
@@ -54,6 +54,107 @@ public class KeymanPackage {
...
+    /**
+     * Indicates that the package is publicly distributed but no longer maintained.  Indicates that other packages targetting
+     * the same targets are more favored.
+     */
+    case deprecated
+
+    /**
+     * Indicates that this package receives full support and maintenance from the resource development community at large.
+     */
+    case publiclyReleased = "publicly released"
```
Comment (mcdurdin) [inline ios/engine/KMEI/KeymanEngine/Classes/KeymanPackage.swift:76]: I'm not very comfortable with this designation and its comment (and thus also `deprecated` and `custom`). It implies a little more knowledge than we can reasonably assume from the package-version query. I think instead, we need to view this a little less in terms of "support state" and more in terms of "mode of distribution":

* `unknown`: we don't know if it's on cloud or not
* `cloud-deprecated`: distributed via Keyman Cloud, but deprecated
* `cloud`: distributed via Keyman Cloud and not deprecated (but we cannot infer support level etc from that)
* `custom`: not distributed via Keyman Cloud (aka ad-hoc, peer-to-peer). Doesn't infer whether or not it is publicly distributed as we cannot determine that.

This suggests that `SupportState` is probably the wrong name for the enum. Maybe instead `DistributionState`?

File: ios/engine/KMEI/KeymanEngine/Classes/KeymanPackage.swift [hunk 2]
```diff
@@ -54,6 +54,107 @@ public class KeymanPackage {
...
+    /**
+     * There is a publicly-distributed version of the package that is newer than the one currently installed within KeymanEngine.
+     */
+    case needsUpdate
+  }
+
+  public enum InstallationState: String, Codable {
+    /**
+     * While a `KeymanPackage` instance will never return it, some functions that accept a `KeymanPackage.Key` might.
+     * This indicates that the package is neither installed nor being downloaded.
```
Comment (mcdurdin) [inline ios/engine/KMEI/KeymanEngine/Classes/KeymanPackage.swift:105]:
```suggestion
     * This indicates that the package is neither installed nor being downloaded.
     * A `KeymanPackage` instance will never return this state, some functions that accept a `KeymanPackage.Key` might.
```
(Which functions might?)

File: ios/engine/KMEI/KeymanEngine/Classes/KeymanPackage.swift [hunk 3]
```diff
@@ -54,6 +54,107 @@ public class KeymanPackage {
...
+     */
+     case none
+
+    /**
+     * While a `KeymanPackage` instance will never return it, some functions that accept a `KeymanPackage.Key` might.
+     * This indicates that the package is not yet installed or downloaded, but is _being_ downloaded via KeymanEngine's downloading queue.
```
Comment (mcdurdin) [inline ios/engine/KMEI/KeymanEngine/Classes/KeymanPackage.swift:111]:
```suggestion
     * This indicates that the package is not yet installed or downloaded, but is _being_ downloaded via KeymanEngine's downloading queue.
     * A `KeymanPackage` instance will never return this state. Some functions that accept a `KeymanPackage.Key` might.
```
(which functions might?)

File: ios/engine/KMEI/KeymanEngine/Classes/KeymanPackage.swift [hunk 4]
```diff
@@ -54,6 +54,107 @@ public class KeymanPackage {
...
+     */
+     case downloading
+
+     /**
+      * Indicates that this `KeymanPackage` instance was been temporarily extracted for installation from a KMP file, regardless
+      * of whether or not its contents have already been installed within KeymanEngine.
```
Comment (mcdurdin) [inline ios/engine/KMEI/KeymanEngine/Classes/KeymanPackage.swift:117]:
```suggestion
      * Indicates that this `KeymanPackage` instance has been temporarily extracted for installation from a KMP file, regardless
      * of whether or not its contents have already been installed within KeymanEngine.
```

File: ios/engine/KMEI/KeymanEngineTests/resources/Queries.bundle/package-version-km-updated.json [hunk 5]
```diff
@@ -0,0 +1,8 @@
+{
+    "keyboards": {
+        "khmer_angkor": {
+            "version": "1.1.0",
+            "kmp": "https://downloads.keyman.com/keyboards/khmer_angkor/1.0.6/khmer_angkor.kmp"
```
Comment (mcdurdin) [inline ios/engine/KMEI/KeymanEngineTests/resources/Queries.bundle/package-version-km-updated.json:5]: This does not match what I'd expect given the version string:
```suggestion
            "kmp": "https://downloads.keyman.com/keyboards/khmer_angkor/1.1.0/khmer_angkor.kmp"
```

**Comment by jahorton** [comment]
> 
> 
> LGTM So the changes I am requesting are really around nomenclature rather than functionality 😁

That's pretty reasonable.  I had to make a "best guess" while designing it, and your arguments regarding nomenclature make sense.  Glad that it's just a nomenclature issue, rather than a design issue.

Took me a little while to reach that design - the old `KeyboardState` enum kind of conflated "install state" and "version state", and I feel like it's a major improvement to disentangle them as this PR does, especially since we're now adding "distribution state."

**Pull request merged**